### PR TITLE
chore(clippy): Fix clippy cast and closure warnings

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Clippy nightly has new warnings, they clean up our code a bit.

## Review

This is a low priority, it doesn't matter if it gets in the release candidate or not.

### Reviewer Checklist

  - [x] Will the PR name make sense to users?
    - [x] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [x] Are the PR labels correct?
  - [x] Does the code do what the ticket and PR says?
  - [ ] How do you know it works? Does it have tests?

### CODE REVIEW

**Review by upbqdn — COMMENTED** [review]
`clippy` on `stable` has some suggestions regarding the `.ok_or_else`.

**Comment by teor2345** [comment]
> `clippy` on `stable` has some suggestions regarding the `.ok_or_else`.

I think stable clippy is wrong, I'll push some ignores when I get a chance.

File: zebra-chain/src/orchard/shielded_data.rs [hunk 1]
```diff
@@ -250,6 +250,6 @@ impl ZcashDeserialize for Flags {
         // the reserved bits 2..7 of the flagsOrchard field MUST be zero."
         // https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus
         Flags::from_bits(reader.read_u8()?)
```
Comment (teor2345) [inline zebra-chain/src/orchard/shielded_data.rs:255]:
```suggestion
        //
        // Clippy 1.64 is wrong here, this lazy evaluation is necessary, constructors are functions. This is fixed in 1.66.
        #[allow(clippy::unnecessary-lazy-evaluations)]
        Flags::from_bits(reader.read_u8()?)
```

File: zebra-chain/src/transaction/serialize.rs [hunk 2]
```diff
@@ -874,9 +874,11 @@ impl ZcashDeserialize for Transaction {
                 // Convert it to a NetworkUpgrade
                 let network_upgrade =
```
Comment (teor2345) [inline zebra-chain/src/transaction/serialize.rs:878]:
```suggestion
                //
                // Clippy 1.64 is wrong here, this lazy evaluation is necessary, constructors are functions. This is fixed in 1.66.
                #[allow(clippy::unnecessary-lazy-evaluations)]
                let network_upgrade =
```

File: zebra-chain/src/orchard/shielded_data.rs [hunk 3]
```diff
@@ -249,7 +249,10 @@ impl ZcashDeserialize for Flags {
         // Consensus rule: "In a version 5 transaction,
         // the reserved bits 2..7 of the flagsOrchard field MUST be zero."
         // https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus
+        //
+        // Clippy 1.64 is wrong here, this lazy evaluation is necessary, constructors are functions. This is fixed in 1.66.
+        #[allow(clippy::unnecessary-lazy-evaluations)]
```
Comment (upbqdn) [inline zebra-chain/src/orchard/shielded_data.rs]:
```suggestion
        #[allow(clippy::unnecessary_lazy_evaluations)]
```

File: zebra-chain/src/transaction/serialize.rs [hunk 4]
```diff
@@ -872,11 +872,16 @@ impl ZcashDeserialize for Transaction {
                 }
                 // Denoted as `nConsensusBranchId` in the spec.
                 // Convert it to a NetworkUpgrade
+                //
+                // Clippy 1.64 is wrong here, this lazy evaluation is necessary, constructors are functions. This is fixed in 1.66.
+                #[allow(clippy::unnecessary-lazy-evaluations)]
```
Comment (upbqdn) [inline zebra-chain/src/transaction/serialize.rs]:
```suggestion
                #[allow(clippy::unnecessary_lazy_evaluations)]
```

**Review by upbqdn — DISMISSED** [review]

File: zebra-chain/src/orchard/shielded_data.rs (hunk 3 above)
Comment (upbqdn) [reply zebra-chain/src/orchard/shielded_data.rs]: We have to use underscores instead of the hyphen-minus.

**Review by upbqdn — COMMENTED** [review]
We have more `clippy::unnecessary-lazy-evaluations`.

**Comment by upbqdn** [comment]
Now I noticed `clippy`'s output uses the hyphen-minus in lint names, but one has to use underscores for lint names in the code. Insidious!

**Comment by teor2345** [comment]
> We have more `clippy::unnecessary-lazy-evaluations`.

Should be fixed now, I put it on the correct line.

**Comment by teor2345** [comment]
> ERROR: (gcloud.compute.ssh) Could not fetch resource:
> - Authentication backend unavailable.

https://github.com/ZcashFoundation/zebra/actions/runs/3238406056/jobs/5307121094#step:6:105

**Review by upbqdn — APPROVED** [review]
:tada:

**Pull request merged**